types.go

```diff
@@ -121,6 +122,18 @@ type functionInfo struct {
 	statuser         Statuser
 }
 
+// TODO(btracey): Think about making this an exported function when the
+// constraint interface is designed.
+func (f functionInfo) satisfies(method Method) error {
+	gradOk := f.IsGradient || f.IsFunctionGradient
+	if method.Needs().Gradient && !gradOk {
+		return errors.New("function does not implement Gradient or FunctionGradient")
```

Comment:
Add the prefix: "optimize: function does not implement needed Gradient or FunctionGradient"